sql/core/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala

```diff
@@ -129,3 +135,11 @@ case class UserDefinedFunction protected[sql] (
     }
   }
 }
+
+object UserDefinedFunction {
+  // This is to keep backward compatibility for this case class.
+  // TODO: revisit this case class in Spark 3.0, and narrow down the public surface.
+  def unapply(arg: UserDefinedFunction): Option[(AnyRef, DataType, Option[Seq[DataType]])] = {
```

Comment:
Doesn't this still break binary compatibility since we bind to another signature?

Reply:
I checked locally and apparently not - looks explicit unapply here and unapply from case class are handled as the same signature from my cursory test.